---
title: "Class15: Genome Annotation and the Interpretation of Gene Lists"
output: github_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

### About our Input Data

The data for for hands-on session comes from GEO entry: GSE37704, which is associated with the following publication:  
  * Trapnell C, Hendrickson DG, Sauvageau M, Goff L et al. "Differential analysis of gene regulation at transcript resolution with RNA-seq". Nat Biotechnol 2013 Jan;31(1):46-53.
  
# Section 1. Differential Expression Analysis
```{r}
# Load our data files
metaFile <- "GSE37704_metadata.csv"
countFile <- "GSE37704_featurecounts.csv"

# Import metadata and take a peak
colData = read.csv(metaFile, row.names=1)
head(colData)
```

Import countdata
```{r}
countData = read.csv(countFile, row.names=1)
head(countData)
```

Remove the troublesome first column from countData
```{r}
# Note we need to remove the odd first $length col

## By using '-1' you include everything EXCEPT the first column
countData <- as.matrix(countData[,-1])
head(countData)
```

Double check that the colnames in countData match the id values in the colData metadata file.

```{r}
colnames(countData)
```
```{r}
rownames(countData)
```

Compare the colnames in countData to the id values int he colData metadata file
```{r}
# Note: using the all() function returns either TRUE/FALSE if all the entries in a vector have the same value
all(colnames(countData) == rownames(colData))
```

Now, remove the genes from countData that have 0 read counts
```{r}
# Filter count data where you have 0 read count across all samples.
countData = countData[rowSums(countData) != 0, ]
head(countData)
```

## Running DESeq2
Nice now lets setup the DESeqDataSet object required for the DESeq() function and then run the DESeq pipeline. This is again similar to our last days hands-on session.

```{r}
library(DESeq2)
```

```{r}
# Setup the object with our data in the war DESeq2 wants it
dds = DESeqDataSetFromMatrix(countData=countData,
                             colData=colData,
                             design=~condition)
# Run the analysis
dds = DESeq(dds)
```

get out results
```{r}
res = results(dds)
```

```{r}
res
```

## Volcano plot
```{r}
# Make a color vector for all genes
mycols <- rep("gray", nrow(res) )

# Color red the genes with absolute fold change above 2
mycols[ abs(res$log2FoldChange) > 2 ] <- "red"

# Color blue those with adjusted p-value less than 0.01
#  and absolute fold change more than 2
inds <- (abs(res$pvalue) < 0.01) & (abs(res$log2FoldChange) > 2 )
mycols[ inds ] <- "blue"

plot( res$log2FoldChange, -log(res$padj), col= c("gray", "red", "blue"), xlab="Log2(FoldChange)", ylab="-Log(P-value)" )
```

## Add gene symbols and entrez ids
```{r}
library("AnnotationDbi")
library("org.Hs.eg.db")
```

We can translate between all the following database ID systems
```{r}
columns(org.Hs.eg.db)
```

Add SYMBOL annotation to our results
```{r}
res$symbol = mapIds(org.Hs.eg.db,
                    keys=row.names(countData),
                    keytype="ENSEMBL",
                    column="SYMBOL",
                    multiVals="first")
res
```

Add ENTREZID annotation to our results
```{r}
res$entrez = mapIds(org.Hs.eg.db,
                    keys=row.names(countData),
                    keytype="ENSEMBL",
                    column="ENTREZID",
                    multiVals="first")
```

Add GENENAME annotation to our results
```{r}
res$name =   mapIds(org.Hs.eg.db,
                    keys=row.names(res),
                    keytype="ENSEMBL",
                    column="GENENAME",
                    multiVals="first")
```

```{r}
head(res,10)
```

# Section 2. Pathway Analysis

Load the packages and setup the KEGG data-sets we need.
```{r}
library(pathview)
```

```{r}
library(gage)
library(gageData)

data(kegg.sets.hs)
data(sigmet.idx.hs)

# Focus on signaling and metabolic pathways only
kegg.sets.hs = kegg.sets.hs[sigmet.idx.hs]

# Examine the first 3 pathways
head(kegg.sets.hs, 3)
```

Note: The main gage() function requires a names vector of fold changes, where the names of the values are the Entrez gene IDs

Note that we used the mapIDs() function above to obtain Entrez gene IDs (stored in res$entrez) and we have the fold change results from DESeq2 analysis (stored in res$log2FoldChange).
```{r}
foldchanges = res$log2FoldChange
names(foldchanges) = res$entrez
head(foldchanges)
```

Now, let's run the **gage** pathway analysis
```{r}
# Get the results
keggres = gage(foldchanges, gsets=kegg.sets.hs)
```

Now, lets look at the object returned from **gage()**
```{r}
attributes(keggres)
```

```{r}
# Look at the first few down (less) pathways
head(keggres$less)
```

Now, let's try out the **pathview()** function from the pathview package to make a pathway plot with our RNA-Seq expression results shown in color.
```{r}
pathview(gene.data=foldchanges, pathway.id="hsa04110")
```

You can play with the other input arguments to pathview() to change the display in various ways including generating a PDF graph. For example:
```{r}
# A different PDF based output of the same data
pathview(gene.data=foldchanges, pathway.id="hsa04110", kegg.native=FALSE)
```





