---
title: "Untitled"
output: github_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```
# Hands-on Section Worksheet Part 1
## 1. The NCI Genomic Data Commons

### Q1  
How many Cases (i.e. patient samples) have been found to have p53 mutations?  
4,008

### Q2  
What are the top 6 misssense mutations found in this gene?  chr17:g.7675088C>T, chr17:g.7673803G>A, chr17:g.7674220C>T, chr17:g.7673802C>T, chr17:g.7674221G>A, chr17:g.7673776G>A

### Q3
Which domain of the protein (as annotated by PFAM) do these mutations reside in?  

### Q4
What are the top 6 primary sites (i.e. cancer locations such as Lung, Brain, etc.) with p53 mutations and how many primary sites have p53 mutations been found in?  

## 2. The GenomicDataCommons R package

Once installed, load the packages, as usual.
```{r}
library(GenomicDataCommons)
library(TCGAbiolinks)
library(maftools)
```

Now let's check on GDC status:
```{r}
status()
```

## 3. Querying the GDC from R

```{r}
projects <- getGDCprojects()
head(projects)
```

You can see all the project names along with their project IDs and associated information.
```{r}
View(projects)
```

Moving onto **cases()** we can use an example from the package associated publication to answer our first from question above:
```{r}
cases_by_project <- cases() %>%
  facet("project.project_id") %>%
  aggregations()
head(cases_by_project)
```

Note that the **facet()** and **aggregations()** functions here are from the GenomicDataCommons package and act to group all cases by the project id and then count them up.

Let's make a barplot of the cases per project. Let's plot this data with a log scale for the y axis, rotated axis labels, and color the bar corresponding to the TCGA-PAAD project.

```{r}
x <- cases_by_project$project.project_id

# Make a custom color vector for our plot
colvec <- rep("lightblue", nrow(x))
colvec[x$key == "TCGA-PAAD"] <- "red"

# Plot with 'log' for y axis and rotate labels with 'las'
par(mar = c(12, 4, 0.5, 3))  
barplot(x$doc_count, names.arg=x$key, log="y", col=colvec, las=2)
```

We can use the **getSampleFilesSummary()** function to determine for a given project how many cases and what type of data we have available for each case:
```{r}
samp <- getSampleFilesSummary("TCGA-PAAD")
head(samp)
```

# Part 2: Designing a Personalized Cancer Vaccine

## Q1: Identify sequence regions that contain all 9-mer peptides that are only found in the tumor.
```{r}
library(bio3d)
seq <- read.fasta("lecture18_sequences.fa")
seq
```












